src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs

```diff
@@ -250,19 +249,25 @@ public static async Task<PackageSpec> GetProjectSpec(IDependencyGraphProject pro
 
                         if (!string.IsNullOrEmpty(outputPath))
                         {
-                            var persistedDGSpecPath = Path.Combine(outputPath, dgFileName);
-
-                            if (File.Exists(persistedDGSpecPath))
+                            for (int frameworkCount = 0; frameworkCount < packageSpec.RestoreMetadata.TargetFrameworks.Count; frameworkCount++)
                             {
-                                var persistedDGSpec = DependencyGraphSpec.Load(persistedDGSpecPath);
-
-                                foreach (var dependentPackageSpec in persistedDGSpec.GetClosure(packageSpec.RestoreMetadata.ProjectUniqueName))
+                                for (var projectReferenceCount = 0; projectReferenceCount < packageSpec.RestoreMetadata.TargetFrameworks[frameworkCount].ProjectReferences.Count; projectReferenceCount++)
                                 {
-                                    if (!(uniqueProjectDependencies.Contains(dependentPackageSpec.RestoreMetadata.ProjectPath) ||
-                                        projects.Any(p => stringComparer.Equals(p.MSBuildProjectPath, dependentPackageSpec.RestoreMetadata.ProjectPath))))
+                                    if (!knownProjects.Contains(packageSpec.RestoreMetadata.TargetFrameworks[frameworkCount].ProjectReferences[projectReferenceCount].ProjectPath))
                                     {
-                                        uniqueProjectDependencies.Add(dependentPackageSpec.RestoreMetadata.ProjectPath);
-                                        dgSpec.AddProject(dependentPackageSpec);
+                                        var persistedDGSpecPath = Path.Combine(outputPath, dgFileName);
+                                        if (File.Exists(persistedDGSpecPath))
+                                        {
+                                            var persistedDGSpec = DependencyGraphSpec.Load(persistedDGSpecPath);
```

Comment:
Is the count of these calls(DependencyGraphSpec.Load) the root cause for the performance degradation? If yes, would be useful to add telemetry that will allow understanding the count of these method calls (maybe performance) per solution?

Reply:
> Is the count of these calls(DependencyGraphSpec.Load) the root cause for the performance degradation? If yes, would be useful to add telemetry that will allow understanding the count of these method calls (maybe performance) per solution? (in case that adding telemetry is possible)

The count was the problem.
Ideally this is only called when a project that used to be in the solution was unloaded. 

We could measure that number specifically but that's probably too detailed at this point. 
Measuring how long the whole operation takes should be enough as a first step. 
Disk hits are more expensive than all other operations, so we should be able to observe it in the telemetry.